File: CoreTransformer/InitialConditionsDistributionCollection.vb
'************************************************************************************
' ST-Sim: A .NET class library for developing state-and-transition simulation models.
'
' Copyright © 2009-2015 ApexRMS.
'
'************************************************************************************

Imports SyncroSim.StochasticTime
Imports System.Collections.ObjectModel

''' <summary>
''' Initial Conditions Distirbution collection
''' </summary>
Friend Class InitialConditionsDistributionCollection
    Inherits Collection(Of InitialConditionsDistribution)

    ''' <summary>
    ''' Get a collection of InitialConditionDistribution objects for the specified Iteration
    ''' </summary>
    ''' <param name="iteration"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function GetForIteration(iteration As Nullable(Of Integer)) As InitialConditionsDistributionCollection
        Dim icds As New InitialConditionsDistributionCollection
        For Each icd In Me
            If (Nullable.Equals(icd.Iteration, iteration)) Then
                icds.Add(icd)
            End If
        Next

        Return icds
    End Function

    ''' <summary>
    ''' Get a Sorted List of Iterations contained in this collection
    ''' </summary>
    ''' <returns>A list of iterations</returns>
    ''' <remarks></remarks>
    Public Function GetSortedIterationList() As List(Of Nullable(Of Integer))

        Dim lstIterations As New List(Of Nullable(Of Integer))
        For Each icd As InitialConditionsDistribution In Me
            Dim iteration = icd.Iteration
            If Not lstIterations.Contains(iteration) Then
                lstIterations.Add(iteration)
            End If
        Next

        'Sort Ascending with Null at start
        lstIterations.Sort()

        Return lstIterations

    End Function

    ''' <summary>
    ''' Get a Filtered Collection of InitialConditionsDistribution for specified parameter
    ''' </summary>
    ''' <returns>A Collection of InitialConditionsDistribution</returns>
    ''' <remarks></remarks>
    Public Function GetFiltered(cell As Cell) As InitialConditionsDistributionCollection

        Dim retVal As New InitialConditionsDistributionCollection()

        For Each icd As InitialConditionsDistribution In Me

            If cell.StratumId <> icd.StratumId Then
                Continue For
            End If

            If cell.StateClassId <> ApexRaster.DEFAULT_NO_DATA_VALUE Then
                If cell.StateClassId <> icd.StateClassId Then
                    Continue For
                End If
            End If

            If cell.SecondaryStratumId <> ApexRaster.DEFAULT_NO_DATA_VALUE Then
                If cell.SecondaryStratumId <> icd.SecondaryStratumId Then
                    Continue For
                End If
            End If

            If cell.Age <> ApexRaster.DEFAULT_NO_DATA_VALUE Then
                If cell.Age < icd.AgeMin Or cell.Age > icd.AgeMax Then
                    Continue For
                End If
            End If

            ' Passed all the tests, so we'll take this one
            retVal.Add(icd)

        Next

        Return retVal

    End Function

    Public Function CalcSumOfRelativeAmount() As Double

        Dim sumOfRelativeAmount As Double = 0.0

        For Each sis As InitialConditionsDistribution In Me
            sumOfRelativeAmount += sis.RelativeAmount
        Next

        Return sumOfRelativeAmount

    End Function

End Class

File: src/Runtime/InitialConditionsDistributionCollection.cs
// stsim: A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2019 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using System;
using SyncroSim.StochasticTime;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace SyncroSim.STSim
{
    /// <summary>
    /// Initial Conditions Distirbution collection
    /// </summary>
    internal class InitialConditionsDistributionCollection : Collection<InitialConditionsDistribution>
    {
        /// <summary>
        /// Get a collection of InitialConditionDistribution objects for the specified Iteration
        /// </summary>
        /// <param name="iteration"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        public InitialConditionsDistributionCollection GetForIteration(int? iteration)
        {
            InitialConditionsDistributionCollection icds = new InitialConditionsDistributionCollection();

            foreach (InitialConditionsDistribution icd in this)
            {
                if (Nullable.Equals(icd.Iteration, iteration))
                {
                    icds.Add(icd);
                }
            }

            return icds;
        }

        /// <summary>
        /// Get a Sorted List of Iterations contained in this collection
        /// </summary>
        /// <returns>A list of iterations</returns>
        /// <remarks></remarks>
        public List<int?> GetSortedIterationList()
        {
            List<int?> Iterations = new List<int?>();

            foreach (InitialConditionsDistribution icd in this)
            {
                var iteration = icd.Iteration;

                if (!Iterations.Contains(iteration))
                {
                    Iterations.Add(iteration);
                }
            }

            //Sort Ascending with Null at start
            Iterations.Sort();

            return Iterations;
        }

        /// <summary>
        /// Get a Filtered Collection of InitialConditionsDistribution for specified parameter
        /// </summary>
        /// <returns>A Collection of InitialConditionsDistribution</returns>
        /// <remarks></remarks>
        public InitialConditionsDistributionCollection GetFiltered(Cell cell)
        {
            InitialConditionsDistributionCollection ICDCollection = new InitialConditionsDistributionCollection();

            foreach (InitialConditionsDistribution icd in this)
            {
                if (cell.StratumId != icd.StratumId)
                {
                    continue;
                }

                if (cell.StateClassId != Spatial.DefaultNoDataValue)
                {
                    if (cell.StateClassId != icd.StateClassId)
                    {
                        continue;
                    }
                }

                if (cell.SecondaryStratumId != Spatial.DefaultNoDataValue)
                {
                    if (cell.SecondaryStratumId != icd.SecondaryStratumId)
                    {
                        continue;
                    }
                }

                if (cell.TertiaryStratumId != Spatial.DefaultNoDataValue)
                {
                    if (cell.TertiaryStratumId != icd.TertiaryStratumId)
                    {
                        continue;
                    }
                }

                if (cell.Age != Spatial.DefaultNoDataValue)
                {
                    if (cell.Age < icd.AgeMin || cell.Age > icd.AgeMax)
                    {
                        continue;
                    }
                }

                // Passed all the tests, so we'll take this one
                ICDCollection.Add(icd);
            }

            return ICDCollection;
        }

        public double CalcSumOfRelativeAmount()
        {
            double sumOfRelativeAmount = 0.0;

            foreach (InitialConditionsDistribution sis in this)
            {
                sumOfRelativeAmount += sis.RelativeAmount;
            }

            return sumOfRelativeAmount;
        }
    }
}
